coalib/bears/Bear.py

```diff
@@ -368,15 +370,35 @@ def download_cached_file(self, url, filename):
         :param filename: The filename it should get, e.g. "test.txt".
         :return:         A full path to the file ready for you to use!
         """
-        filename = join(self.data_dir, filename)
-        if exists(filename):
-            return filename
+        scrape = requests.head(url)
 
+        filename = join(self.data_dir, filename)
+        versionfile = join(self.data_dir, filename + 'version.txt')
+        if exists(versionfile):
+            with open(versionfile, 'r') as vfile:
+                etag = vfile.readline().rstrip('\n')
+                last_modified = vfile.readline().rstrip('\n')
+            if exists(filename):
+                if 'etag' in scrape.headers:
+                    if scrape.headers['etag'] == etag:
+                        return filename
+                elif 'last-modified' in scrape.headers:
+                    if scrape.headers['last-modified'] == last_modified:
+                        return filename
+
+        with open(versionfile, 'w') as vfile:
+            if 'etag' in scrape.headers:
+                vfile.write(scrape.headers['etag'] + '\n')
+            else:
+                vfile.write('\n')
+            if 'last-modified' in scrape.headers:
+                vfile.write(scrape.headers['last-modified'] + '\n')
         self.info('Downloading {filename!r} for bear {bearname} from {url}.'
                   .format(filename=filename, bearname=self.name, url=url))
-
-        with urlopen(url) as response, open(filename, 'wb') as out_file:
-            copyfileobj(response, out_file)
+        download = requests.get(url, stream=True)
```

Comment:
this means ftp content cant be used, as urllib supports ftp:// and file://, but requests does not.

docstring needs to indicate this